### PR TITLE
Speedup the way user accounts are funded

### PR DESCRIPTION
**Pull request by sparrowDom**
- Fund user accounts via storage slot manipulation rather than fetching tokens from other accounts
- make it possible to change the price of any chainlink oracle

### CODE REVIEW

File: contracts/deploy/999_fork_test_setup.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ const {
 const { fundAccounts } = require("../utils/funding");
 const addresses = require("../utils/addresses");
 const daiAbi = require("../test/abi/dai.json").abi;
+const { hardhatSetBalance } = require("../test/_fund");
```
Comment (naddison36) [inline contracts/deploy/999_fork_test_setup.js:10]: I also moved `hardhatSetBalance` into a util file on one of my branches. I can manage to consolidate it when I merge this back into my branches

File: contracts/package.json [hunk 2]
```diff
@@ -38,6 +38,7 @@
   "author": "Origin Protocol Inc <support@originprotocol.com>",
   "license": "MIT",
   "devDependencies": {
+    "@nomicfoundation/hardhat-network-helpers": "^1.0.9",
```
Comment (naddison36) [inline contracts/package.json:41]: nice addition. We can start following Hardhat's recommendation and migrate away from Waffle with this plug-in https://hardhat.org/hardhat-runner/docs/guides/migrating-from-hardhat-waffle

File: contracts/package.json [hunk 3]
```diff
@@ -58,6 +59,7 @@
     "hardhat-deploy": "^0.11.30",
     "hardhat-deploy-ethers": "^0.3.0-beta.13",
     "hardhat-gas-reporter": "^1.0.9",
+    "hardhat-network-helpers": "^0.0.1",
```
Comment (naddison36) [inline contracts/package.json:62]: Isn't this a duplicate of `@nomicfoundation/hardhat-network-helpers`?

**Review by naddison36 — CHANGES_REQUESTED** [review]
There's a number of lint errors that need fixing

File: contracts/test/helpers.js [hunk 4]
```diff
@@ -152,15 +152,6 @@ chai.Assertion.addMethod("emittedEvent", async function (eventName, args) {
   }
 });
 
-/**
- * Converts an amount in the base unit of a contract to the standard decimal unit for the contract.
- * @param {string} amount - The amount to convert, represented as a string in the base unit of the contract.
- * @param {Contract} contract - The token contract to get the decimal places for.
- */
-async function units(amount, contract) {
```
Comment (naddison36) [inline contracts/test/helpers.js:160]: nice removal of duplicate functions

File: contracts/test/_fund.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+  for (let i = 0; i < 100; i += 1) {
+    const probedSlot = keccak256(encode(["address", "uint"], [account, i]));
+    if (probedSlot.startsWith("0x0")) continue;
+    // remove padding for JSON RPC
+    // while (probedSlot.startsWith("0x0")) {
+    //     probedSlot = `0x${probedSlot.slice(3)}`
+    // }
+
+    const prev = await network.provider.send("eth_getStorageAt", [
```
Comment (naddison36) [inline contracts/test/_fund.js:44]: @nomicfoundation/hardhat-network-helpers has a `getStorageAt` function

File: contracts/test/_fund.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    const probe = prev === probeA ? probeB : probeA;
+
+    await setStorageAt(tokenAddress, probedSlot, probe);
+
+    const balance = await token.balanceOf(account);
+
+    // reset to previous value
+    await setStorageAt(tokenAddress, probedSlot, prev);
+
+    if (balance.eq(ethrs.BigNumber.from(probe))) return i;
```
Comment (naddison36) [inline contracts/test/_fund.js:59]: it'd be nice to cache the results so we don't need to keep finding the slots of the tokens we have already set balances for

File: contracts/deploy/999_fork_test_setup.js (hunk 1 above)
Comment (sparrowDom) [reply contracts/deploy/999_fork_test_setup.js:10]: thanks! whoever merges first wins :)

File: contracts/package.json (hunk 2 above)
Comment (sparrowDom) [reply contracts/package.json:41]: Have tried quickly doing it by following the migration guide but got an error: 
Screenshot 2023-10-24 at 16 45 10

have created a [Notion task](https://www.notion.so/originprotocol/Migrate-from-Waffle-to-hardhat-27c543b54a91421b8bbd2e2f1f4be617
) for it:

File: contracts/package.json (hunk 3 above)
Comment (sparrowDom) [reply contracts/package.json]: Yes, thanks removing

File: contracts/test/_fund.js (hunk 5 above)
Comment (sparrowDom) [reply contracts/test/_fund.js]: Thanks, that cleans up the code nicely. Have also added the addresses of the first 4 accounts and contracts to speed up funding a bit more: https://github.com/OriginProtocol/origin-dollar/pull/1872/commits/7396c2b63aa9c85a57de6d6a7ecad26e6900f51f

File: contracts/test/_fund.js (hunk 6 above)
Comment (sparrowDom) [reply contracts/test/_fund.js]: thanks this is very good point, and a significant speed increase: https://github.com/OriginProtocol/origin-dollar/pull/1872/commits/960914237f5eaa29fef2dd633a71d148f0631c6b

**Comment by sparrowDom** [comment]
@naddison36 do you want to take another look at this PR please?

**Review by naddison36 — APPROVED** [review]
looks great. I'm keen to use this on my feature branches

**Review by shahthepro — APPROVED** [review]

**Pull request merged**